// Copyright (c) 2023 Boston Dynamics, Inc.  All rights reserved.
//
// Downloading, reproducing, distributing or otherwise using the SDK Software
// is subject to the terms and conditions of the Boston Dynamics Software
// Development Kit License (20191101-BDSDK-SL).

syntax = "proto3";

package bosdyn.api;
option java_outer_classname = "GripperCameraParamProto";

import "bosdyn/api/header.proto";
import "bosdyn/api/geometry.proto";
import "google/protobuf/wrappers.proto";

// The GripperCameraParam request message sets new gripper sensor parameters. Gripper sensor
// parameters do not persist across reboots.
message GripperCameraParamRequest {
    // Common request header.
    RequestHeader header = 1;

    GripperCameraParams params = 2;
}

message GripperCameraParamResponse {
    // Common response header.
    ResponseHeader header = 1;
}

// The GripperCameraGetParam request message queries the robot for the current gripper sensor
// parameters.
message GripperCameraGetParamRequest {
    // Common request header.
    RequestHeader header = 1;
}

// The GripperCameraGetParam response message contains the current gripper sensor parameters.
// Gripper sensor parameters do not persist across reboots.
message GripperCameraGetParamResponse {
    // Common request header.
    ResponseHeader header = 1;

    GripperCameraParams params = 2;
}

message GripperCameraParams {
    enum CameraMode {
        option allow_alias = true;

        // MODE_UNKNOWN should not be used.
        MODE_UNKNOWN = 0;

        // 640x480 pixels.
        MODE_640_480 = 11;
        MODE_640_480_120FPS_UYVY = 11 [deprecated = true];

        // 1280x720 pixels.
        MODE_1280_720 = 1;
        MODE_1280_720_60FPS_UYVY = 1 [deprecated = true];

        // 1920x1080 pixels.
        MODE_1920_1080 = 14;
        MODE_1920_1080_60FPS_MJPG = 14 [deprecated = true];

        // 3840x2160 pixels.
        MODE_3840_2160 = 15;
        MODE_3840_2160_30FPS_MJPG = 15 [deprecated = true];

        // 4096x2160 pixels.
        MODE_4096_2160 = 17;
        MODE_4096_2160_30FPS_MJPG = 17 [deprecated = true];

        // 4208x3120 pixels.
        MODE_4208_3120 = 16;
        MODE_4208_3120_20FPS_MJPG = 16 [deprecated = true];

    }

    // CameraMode sets the resolution, frame rate and image format.
    CameraMode camera_mode = 1;

    // Set the image brightness level.
    // Min 0, max 1
    google.protobuf.FloatValue brightness = 2;

    // Set the image contrast level.
    // Min 0, max 1
    google.protobuf.FloatValue contrast = 3;

    // Set the image saturation level.
    // Min 0, max 1
    google.protobuf.FloatValue saturation = 4;

    // Set the image gain level.
    // This paramater is only effective when manual exposure is used.
    // Min 0, max 1
    google.protobuf.FloatValue gain = 7;

    // Whether the camera should use auto exposure.
    // Unset is equivalent to setting exposure_auto = true
    google.protobuf.BoolValue exposure_auto = 10;

    // Manually set the image exposure level. This value is only used if exposure_auto is false.
    // Min 0, max 1
    google.protobuf.FloatValue exposure_absolute = 11;

    // Region of interest for exposure.  Specify a spot exposure on a
    // certain part of the image.  Only used in auto-exposure mode.
    RoiParameters exposure_roi = 16;

    // Whether the camera should automatically focus the image.
    // Unset is equivalent to setting focus_auto = true
    google.protobuf.BoolValue focus_auto = 13;

    // Manually set the image focus. This value is only used if focus_auto is false.
    // If focus_auto is true, this value will be populated with the current focus value.
    // Min 0, max 1
    // 0 corresponds to focus at infinity, 1 corresponds to a focal point close to the camera.
    google.protobuf.FloatValue focus_absolute = 12;

    // Region of interest for focus.  Only used when in auto-focus mode.
    RoiParameters focus_roi = 14;

    // Set to true to draw a rectangle in the image where the focus ROI is.
    // Unset is equivalent to setting draw_focus_roi_rectangle = false
    google.protobuf.BoolValue draw_focus_roi_rectangle = 18;

    // High dynamic range (HDR) mode sets the camera to take multiple frames to get exposure
    // in a large range.  HDR will reduce framerate in high-framerate modes.
    HdrParameters hdr = 17;

    enum LedMode {
        // LED_MODE_UNKNOWN should not be used.
        LED_MODE_UNKNOWN = 0;

        // Off
        LED_MODE_OFF = 1;

        // Constantly on. Brightness level can be set in the led_torch_brightness field.
        LED_MODE_TORCH = 2;
    }
    // Set the LED mode.
    LedMode led_mode = 19;

    // Brightness of the LED in torch mode.  Min = 0, max = 1.
    // Note: A brightness value of 0 is *not* off, but is the minimum brightness.
    //       To turn off the LED, set the led_mode to LED_MODE_OFF
    google.protobuf.FloatValue led_torch_brightness = 20;

    // Whether the camera should use auto white balance
    // Unset is equivalent to setting white_balance_temperature_auto = true
    google.protobuf.BoolValue white_balance_temperature_auto = 5;

    // Set the image gamma level.
    // Min 0, max 1
    google.protobuf.FloatValue gamma = 6;

    // Manually set the white balance focus. This value is only used if white_balance_temperature_auto is false.
    // Min 0, max 1
    // 0 corresponds to focus at infinity, 1 corresponds to a focal point close to the camera.
    google.protobuf.FloatValue white_balance_temperature = 8;
    
    // Set the image sharpness level.
    // Min 0, max 1
    google.protobuf.FloatValue sharpness = 9;

    
}

// High dynamic range (HDR) modes available. HDR sets the camera to take multiple frames to
// get exposure in a large range.  HDR will reduce framerate in high-framerate modes.
enum HdrParameters {
    HDR_UNKNOWN = 0;   // (or not set): will not change HDR settings.
    HDR_OFF = 1;       // HDR disabled
    HDR_AUTO = 2;      // Camera's on-board processor determines how much HDR is needed
    HDR_MANUAL_1 = 3;  // Manual HDR enabled (minimum)
    HDR_MANUAL_2 = 4;  //
    HDR_MANUAL_3 = 5;  //
    HDR_MANUAL_4 = 6;  // Manual HDR enabled (maximum)
}

// Region of interest (ROI) indicates the region within the image that should be used for
// determination of automatic focus or exposure.
message RoiParameters {
    // Center point of the ROI in the image. The upper lefthand corner of the image is (0, 0) and
    // the lower righthand corner is (1, 1). The middle of the image is (0.5, 0.5).
    Vec2 roi_percentage_in_image = 1;

    enum RoiWindowSize {
        // ROI window size, 1 is the smallest, 8 is the largest.
        ROI_WINDOW_SIZE_UNKNOWN = 0;
        ROI_WINDOW_SIZE_1 = 1;
        ROI_WINDOW_SIZE_2 = 2;
        ROI_WINDOW_SIZE_3 = 3;
        ROI_WINDOW_SIZE_4 = 4;
        ROI_WINDOW_SIZE_5 = 5;
        ROI_WINDOW_SIZE_6 = 6;
        ROI_WINDOW_SIZE_7 = 7;
        ROI_WINDOW_SIZE_8 = 8;
    }

    // Size of the region of interest.
    RoiWindowSize window_size = 2;
}
